### PR TITLE
Fixed Temperature Scaling

### PR DESCRIPTION
**Pull request by dangoh123**
Have managed to fix the temperature values, using a mass based rescaling factor after the the values have been generated, for more details see commit `9dbefd2`. 

One issue we noticed that @fjebaker might be able to provide some insight into is the optimisation, as in order to generate this scaling factor, we have to use a second value function, which is simply for radius, but this almost doubles the render time. I'm sure there is probably a way to extract the radius values without repeating the geodesic calculations, but we don't really have a firm enough grip on the `ValueFunction` and how that works to be able to apply that.

### CODE REVIEW

**Comment by fjebaker** [comment]
If you're just multiplying the temperature by a value that depends on the radius, you could implement a single value function, along the line of 

```julia
# ... 

M_phys = mass*1.99e30
r_isco_phys = AccretionFormulae.r_isco(M_phys, 0.998)
r_g_phys = M_phys

# ... 

function my_function(m, sol, max_time; kwargs...)
    g = AccretionFormulae.redshift(m, sol, max_time; kwargs...)
    u = sol.u[end]
    M = m.M
    a_star = m.a
    temperature = AccretionFormulae.observed_temperature(u[2], a_star, M, g)
    
    # ...
    
    numerator = AccretionFormulae.mass_scale_fraction.(M_phys, η_phys, edd_ratio_phys, r_isco_phys, r_g_phys, u[2]*M_phys)
    denom = AccretionFormulae.mass_scale_fraction.(M, η, edd_ratio, R_isco, M, u[2])
    
    temperature * numerator / denom
end

my_vf = ValueFunction(my_function)

# ... 
```
I haven't tried running your code like this, but you shouldn't need to run a new ValueFunction to get _another_ value. They are designed to be composable, so this should work fine for you?

I'm currently working on adding render caches and exports though, so you can save geodesics and run mutliple ValueFunctions on them without having to re-render the image. If the above isn't sufficient for whatever reason, by Monday this should be available to you.

If you make changes, please just push them to this PR!

**Comment by dangoh123** [comment]
I had a look at that implementation, and it would work in theory, but with the way our variables are used, it makes it quite awkward to make, so I think we'll stick with our implementation for the time being, until the render caches are ready.

File: scripts/animation_test.jl [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
     print("Frame: $i / $n_frames complete")
-    fangle = Float64(angle)
-    temperature_render(obs_angle=fangle, tolerance = 1e-4)
+    # fangle = Float64(angle)
+    fmass = Float64(mass)
+    temperature_render(mass=mass, tolerance=1e-10, resolution=1080, size_multiplier=4, dtmax=2.5)
 end
-gif(anim, "hello.gif", fps=10)
\ No newline at end of file
+gif(anim, "massgif.gif", fps=20)
```
Comment (fjebaker) [inline scripts/animation_test.jl:13]: Can you post the gif?

File: scripts/histogram_animation.jl [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# vars = 75.0:1.0:85.0                 # angle
+vars = [-179.0:5.0:-1.0;1.0:5.0:179.0] # angle
+# vars = 1.0:5.0:200                   # mass
+n_frames = length(vars)
+anim = @animate for (i, var) in enumerate(vars)
+    print("Frame: $i / $n_frames complete")
+    fvar = Float64(var)
+    energy_histogram(obs_angle=fvar)
+end
+gif(anim, "histgif.gif", fps=20)
```
Comment (fjebaker) [inline scripts/histogram_animation.jl:12]: Also can you post the gif?

File: scripts/histogram_test.jl [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    end
+
+    histogram(lines, 
+            xlims=(0,10),
+            nbins=1000)
+    title!("Observation Angle = $obs_angle")
+    # temperature_render(obs_angle=85.0, mass=1)
+end
+
+energy_histogram()
```
Comment (fjebaker) [inline scripts/histogram_test.jl:150]: Post plot too?

File: scripts/histogram_test.jl [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    redshift_img = @time rendergeodesics(
+        m, u, 2000.0, 
+        d, 
+        fov_factor=fov*size_multiplier, abstol=tolerance, reltol=tolerance,
+        image_width = 350*size_multiplier,
+        image_height = 250*size_multiplier,
+        vf = redshift_vf,
+        dtmax = dtmax
+    )
```
Comment (fjebaker) [inline scripts/histogram_test.jl:91]: Ok, we'll leave this tripple render for now. Should hopefully be able to push the render cache later today, so can tidy this up later.

**Review by fjebaker — APPROVED** [review]
LGTM, we'll look into documenting functions soon so it's clear which ones are unitless (/arbitrary units), and which manifestly return units.

File: scripts/animation_test.jl (hunk 1 above)
Comment (dangoh123) [reply scripts/animation_test.jl:13]: 275733518_1007480233212804_3890997483161396099_n

This one looks pretty good, and we think everything here is physically correct. One small thing would be to centre the black hole on zero, so the size of the disc in gravitational radii (or some other unit?) could be a bit clearer.

File: scripts/histogram_test.jl (hunk 3 above)
Comment (dangoh123) [reply scripts/histogram_test.jl:150]: plot_3

This is just a basic one we made, and doesn't really look right, but some of the information [below](https://github.com/astro-group-bristol/AccretionFormulae.jl/pull/25#issuecomment-1066996047) should help us.

File: scripts/histogram_test.jl (hunk 4 above)
Comment (dangoh123) [reply scripts/histogram_test.jl:91]: Sounds good

**Comment by dangoh123** [comment]
@phajy

**Comment by dangoh123** [comment]
Hi @phajy , we have been able to have a look at this, and made a bit of progress, and might have something that looks promising. For every pixel, we found the emitted flux for the radius of that pixel, then we determined the energy bin based on the redshift of the emitted energy (6.4 keV) and added the flux per unit area in this pixel the corresponding energy bin.

The bins were found in slightly convoluted way [(lines 173-181)](https://github.com/dangoh123/AccretionFormulae.jl/blob/3067c6b4395f0c128e368ae723857c7eaa8d5fd8/scripts/histogram_test.jl), but seems to work for now. The flux per unit area was found by relating the number of pixels the image spanned to the radius of the black hole.

Here is a plot that we got:
plot_47

It looks quite bad, but we noticed that in the area that the lines don't drop down to zero, it seems to resemble the shape we expect, and there could be some sort of noise we have accidentally introduced?

When we isolated the 5-10 keV region, we got:

e_min set to 5

Which could be in the right direction?

**Comment by kyle8mcknight** [comment]
> Hi @phajy , we have been able to have a look at this, and made a bit of progress, and might have something that looks promising. For every pixel, we found the emitted flux for the radius of that pixel, then we determined the energy bin based on the redshift of the emitted energy (6.4 keV) and added the flux per unit area in this pixel the corresponding energy bin.
> 
> The bins were found in slightly convoluted way [(lines 173-181)](https://github.com/dangoh123/AccretionFormulae.jl/blob/3067c6b4395f0c128e368ae723857c7eaa8d5fd8/scripts/histogram_test.jl), but seems to work for now. The flux per unit area was found by relating the number of pixels the image spanned to the radius of the black hole.
> 
> Here is a plot that we got: plot_47
> 
> It looks quite bad, but we noticed that in the area that the lines don't drop down to zero, it seems to resemble the shape we expect, and there could be some sort of noise we have accidentally introduced?
> 
> When we isolated the 5-10 keV region, we got:
> 
> e_min set to 5
> 
> Which could be in the right direction?

Just wanted to add that the units for flux have been clumsily implemented for now but we want to get the shape before anything else - we will revisit in the morning with fresh heads. We'll also take a look at using an arbitrary function for the emissivity profile.

**Comment by phajy** [comment]
Thanks for the code update and example plots. I've made some further updates and will submit a new Pull request.

**Comment by phajy** [comment]
No problem. And sorry again for just pushing the changes to the repository - I forgot that I hadn't "forked" this project so thought I was pushing it to my repository not astro-group-bristol!

To get a decent line profile you need to have enough pixels in the region you're interested in (i.e., the inner disk) to create a smooth spectrum for a given number of spectral bins. You can, of course, play around with r_in, r_out, inclination, and emissivity profile (which ideally should be the Page & Thorne profile from the temperature map).

Finally, you should just be able to do the integration once and have a value function that returns a vector of, e.g., [g, r, temperature] or whatever. That would make execution of the code much faster.

**Comment by kyle8mcknight** [comment]
For sure, that was something we had planned, we're going to look at cleaning up and optimisation very soon.

**Comment by fjebaker** [comment]
@kyle8mcknight the caching method should be ready tonight -- I'm sorry it's been delayed a couple of days. I encountered an issue with type inference for allocating the output array, but I think that is fixed now! I will run the smoke tests to check when I get home and make this available to you soon.

**Comment by fjebaker** [comment]
@phajy @kyle8mcknight @dangoh123 Caching is ready:
- https://github.com/astro-group-bristol/GeodesicRendering.jl/pull/6

**Comment by fjebaker** [comment]
Also, just as a heads up
- https://github.com/astro-group-bristol/GeodesicRendering.jl/issues/8

Pre-rendering like this is heavily memory limited, and is bottle-necked enormously by garbage collection. I'll look into optimizing this very soon.

**Comment by dangoh123** [comment]
Interesting, I'll have a look at this tonight and give you a shout tomorrow if we have any trouble.

**Pull request merged**